### PR TITLE
SAAS-4125: Added App Mesh TLS Required Plugin

### PR DESCRIPTION
**Pull request by sadeed12345**

### CODE REVIEW

File: plugins/aws/appmesh/appmeshTLSRequired.js [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+                    let virtual = describeVirtualGateway.data.virtualGateway;
+                   
+                    if (!virtual.spec.listeners.length) {
+                        helpers.addResult(results, 0,
+                            'App Mesh virtual gateway does not have listeners', region, gatewayArn);
+                    } else {
+                        const tlsEnabled = virtual.spec.listeners.every(listener => listener.tls && listener.tls.mode && listener.tls.mode.toUpperCase() === 'STRICT');
+                        const status = tlsEnabled ? 0 : 2;
+                        helpers.addResult(results, status,
+                            `App Mesh vitual gateway listeners ${tlsEnabled ? 'restrict ' : 'does not restrict '} TLS enabled connections`,
```
Comment (sadeed12345) [inline plugins/aws/appmesh/appmeshTLSRequired.js:84]:
```suggestion
                            `App Mesh virtual gateway listeners ${tlsEnabled ? 'restrict ' : 'does not restrict '} TLS enabled connections`,
```

File: plugins/aws/appmesh/appmeshTLSRequired.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'App Mesh TLS Required',
+    category: 'App Mesh',
+    domain: 'Content Delivery',
+    description: 'Ensure that AWS App Mesh virtual gateways Listener only accepts connections with TLS enabled.',
+    more_info: 'In App Mesh, Transport Layer Security (TLS) encrypts communication between the Envoy proxies deployed on compute resources that are represented in App Mesh by mesh endpoints. An object that represents the Transport Layer Security (TLS) properties for a listener.',
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/appmeshTLSRequired.js:9]: Make it more meaningful

File: plugins/aws/appmesh/appmeshTLSRequired.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+                return rcb();
+            }
+
+            if (!listMeshes.data.length) {
+                helpers.addResult(results, 0, 'No App Mesh meshes found', region);
+                return rcb();
+            }
+
+            for (let mesh of listMeshes.data){
+                if (!mesh.meshName) continue;
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/appmeshTLSRequired.js:37]:
```suggestion
                if (!mesh.meshName || !mesh.arn) continue;
```

File: plugins/aws/appmesh/appmeshTLSRequired.js [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+                if (!mesh.arn) continue;
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/appmeshTLSRequired.js:38]:
```suggestion
```

File: plugins/aws/appmesh/appmeshTLSRequired.js [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+                    var gatewayArn = gateway.arn;
+
+                    var describeVirtualGateway = helpers.addSource(cache, source,
+                        ['appmesh', 'describeVirtualGateway', region, gateway.virtualGatewayName]);
+
+                    if (!describeVirtualGateway ||
+                        describeVirtualGateway.err ||
+                        !describeVirtualGateway.data) {
+                        helpers.addResult(results, 3,
+                            'Unable to App Mesh virtual gateways: ' + helpers.addError(describeVirtualGateway), region, gatewayArn);
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/appmeshTLSRequired.js:71]:
```suggestion
                            'Unable to describe App Mesh virtual gateway: ' + helpers.addError(describeVirtualGateway), region, gatewayArn);
```

File: plugins/aws/appmesh/appmeshTLSRequired.js [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+                if (!listVirtualGateways.data.virtualGateways || !listVirtualGateways.data.virtualGateways.length) {
+                    helpers.addResult(results, 0,
+                        'No App Mesh virtual gateways found', region, resource);
+                    continue;
+                }
+
+                for (let gateway of listVirtualGateways.data.virtualGateways) {
+                    if (!gateway.virtualGatewayName) continue;
+                    if (!gateway.arn) continue;
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/appmeshTLSRequired.js:60]:
```suggestion
                    if (!gateway.virtualGatewayName || !gateway.arn) continue;
```

File: plugins/aws/appmesh/appmeshTLSRequired.js [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+                    var gatewayArn = gateway.arn;
+
+                    var describeVirtualGateway = helpers.addSource(cache, source,
+                        ['appmesh', 'describeVirtualGateway', region, gateway.virtualGatewayName]);
+
+                    if (!describeVirtualGateway ||
+                        describeVirtualGateway.err ||
+                        !describeVirtualGateway.data) {
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/appmeshTLSRequired.js:69]:
```suggestion
                        !describeVirtualGateway.data || describeVirtualGateway.data.virtualGateway) {
```

File: plugins/aws/appmesh/appmeshTLSRequired.js [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+                        helpers.addResult(results, 3,
+                            'Unable to App Mesh virtual gateways: ' + helpers.addError(describeVirtualGateway), region, gatewayArn);
+                        continue;
+                    }
+
+                    let virtual = describeVirtualGateway.data.virtualGateway;
+                   
+                    if (!virtual.spec.listeners.length) {
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/appmeshTLSRequired.js:77]:
```suggestion
                    if (!describeVirtualGateway.data.virtualGateway.spec ||
                        !describeVirtualGateway.data.virtualGateway.spec.listeners ||
                        !describeVirtualGateway.data.virtualGateway.spec.listeners.length) {
```

File: plugins/aws/appmesh/appmeshTLSRequired.js [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+                        helpers.addResult(results, 0,
+                            'App Mesh virtual gateway does not have listeners', region, gatewayArn);
+                    } else {
+                        const tlsEnabled = virtual.spec.listeners.every(listener => listener.tls && listener.tls.mode && listener.tls.mode.toUpperCase() === 'STRICT');
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/appmeshTLSRequired.js:81]:
```suggestion
                        const tlsEnabled = describeVirtualGateway.data.virtualGateway.spec.listeners.every(listener => listener.tls && listener.tls.mode && listener.tls.mode.toUpperCase() === 'STRICT');
```

File: plugins/aws/appmesh/appmeshTLSRequired.spec.js [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+
+describe('appmeshTLSRequired', function () {
+    describe('run', function () {
+        it('should PASS if App Mesh vitual gateway listeners restrict TLS enabled connections', function (done) {
+            const cache = createCache([listMeshes[0]], [listVirtualGateways[0]], describeVirtualGateway[0]);
+            appmeshTLSRequired.run(cache, { }, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(0);
+                expect(results[0].region).to.equal('us-east-1');
+                done();we
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/appmeshTLSRequired.spec.js:213]:
```suggestion
                done();
```

File: plugins/aws/appmesh/appmeshTLSRequired.spec.js [hunk 11]
```diff
@@ -0,0 +1,262 @@
...
+        it('should FAIL if App Mesh vitual gateway listeners does not restrict TLS enabled connections', function (done) {
+            const cache = createCache([listMeshes[1]], [listVirtualGateways[1]], describeVirtualGateway[1]);
+            appmeshTLSRequired.run(cache, { }, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(2);
+                expect(results[0].region).to.equal('us-east-1');
+                done();
+            });
+        });
+
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/appmeshTLSRequired.spec.js:226]: Add message check

**Review by AkhtarAmir — CHANGES_REQUESTED** [review]

File: plugins/aws/appmesh/appmeshTLSRequired.js [hunk 12]
```diff
@@ -0,0 +1,94 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'App Mesh TLS Required',
+    category: 'App Mesh',
+    domain: 'Content Delivery',
+    description: 'Ensure that AWS App Mesh virtual gateways Listener only accepts connections with TLS enabled.',
+    more_info: 'In App Mesh, Transport Layer Security (TLS) encrypts communication between the Envoy proxies deployed on compute resources that are represented in App Mesh by mesh endpoints.',
```
Comment (sadeed12345) [inline plugins/aws/appmesh/appmeshTLSRequired.js:9]:
```suggestion
    more_info: 'In App Mesh, Transport Layer Security (TLS) encrypts communication between the Envoy proxies deployed on compute resources that are represented in App Mesh by mesh endpoints, such as Virtual nodes and Virtual gateways. The proxy negotiates and terminates TLS.',
```

File: plugins/aws/appmesh/appmeshTLSRequired.js [hunk 13]
```diff
@@ -0,0 +1,94 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'App Mesh TLS Required',
+    category: 'App Mesh',
+    domain: 'Content Delivery',
+    description: 'Ensure that AWS App Mesh virtual gateways Listener only accepts connections with TLS enabled.',
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/appmeshTLSRequired.js:8]:
```suggestion
    description: 'Ensure that AWS App Mesh virtual gateway listeners only accepts TLS enabled connections.',
```

File: plugins/aws/appmesh/appmeshTLSRequired.js [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+    more_info: 'In App Mesh, Transport Layer Security (TLS) encrypts communication between the Envoy proxies deployed on compute resources that are represented in App Mesh by mesh endpoints, such as Virtual nodes and Virtual gateways. The proxy negotiates and terminates TLS.',
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/appmeshTLSRequired.js:9]:
```suggestion
    more_info: 'In App Mesh, Transport Layer Security (TLS) encrypts communication between the envoy proxies deployed on compute resources that are represented in App Mesh by mesh endpoints, such as Virtual nodes and Virtual gateways.',
```

File: plugins/aws/appmesh/appmeshTLSRequired.js [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+    link: 'https://docs.aws.amazon.com/app-mesh/latest/APIReference/API_ListenerTls.html',
+    recommended_action: 'Restrict TLS enabled connections to AWS App Mesh virtual gateways Listeners in your account',
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/appmeshTLSRequired.js:11]:
```suggestion
    recommended_action: 'Restrict AWS App Mesh virtual gateway listeners to accept only TLS enabled connections.',
```

**Review by AkhtarAmir — APPROVED** [review]

**Pull request merged**